File: erp-backend/src/main/java/com/erp/common/ConsigneCodes.java
package com.erp.common;

import java.util.Collection;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Registre des codes produits traités comme consigne/emballage, scopé par entreprise (companyId).
 * Chaque entreprise a son propre jeu de codes supplémentaires (issus de sa config eLeader),
 * rechargeable à chaud via refresh(companyId, ...) après sauvegarde de la configuration eLeader.
 */
public final class ConsigneCodes {

    /** Codes de base — toujours présents pour toute entreprise, jamais supprimés. */
    private static final Set<String> BASE = Set.of(
        "CB12", "CB24", "CB12M", "CB24M",
        "CV12", "CV24",
        "CBG12", "CBG15", "CBG24",
        "VIP12", "VIP24", "VCP12", "VCP24",
        "VIPG12", "VIPG15", "VIPG24",
        "CVG12", "CVG15", "CVG24",
        "EGUI12", "EGUI15", "EGUI24",
        "PP", "PB", "TT", "BPM", "BGM",
        "CAIMET", "CONS001", "INPN33",
        "EMB1", "EMB2", "EMB3", "EMB4", "EMB5",
        "CAISMB", "PALT-V", "PALTPL"
    );

    // Thread-safe — per-company set, readable without lock, replaced atomically on refresh
    // for the given company only (other companies' entries untouched).
    private static final ConcurrentHashMap<Long, Set<String>> CODES_BY_COMPANY = new ConcurrentHashMap<>();

    private ConsigneCodes() {}

    public static boolean isConsigne(String productCode, Long companyId) {
        if (productCode == null || productCode.isBlank()) return false;
        Set<String> codes = companyId == null ? null : CODES_BY_COMPANY.get(companyId);
        if (codes == null) codes = BASE;
        return codes.contains(productCode.trim().toUpperCase());
    }

    /**
     * Rechargement à chaud pour une entreprise donnée : BASE + codes supplémentaires
     * issus de la config eLeader de cette entreprise. Appelé au démarrage
     * (ApplicationReadyEvent) et après chaque sauvegarde de config, pour la companyId concernée.
     * Le swap est atomique et n'affecte que l'entrée de cette entreprise dans la map.
     */
    public static void refresh(Long companyId, Collection<String> extraCodes) {
        if (companyId == null) return;
        Set<String> next = ConcurrentHashMap.newKeySet();
        next.addAll(BASE);
        if (extraCodes != null) {
            for (String c : extraCodes) {
                if (c != null && !c.isBlank()) {
                    next.add(c.trim().toUpperCase());
                }
            }
        }
        CODES_BY_COMPANY.put(companyId, next); // atomic per-company swap
    }

    /** Retourne une copie immuable du set actif pour une entreprise (pour affichage/debug). */
    public static Set<String> snapshot(Long companyId) {
        Set<String> codes = companyId == null ? null : CODES_BY_COMPANY.get(companyId);
        return Set.copyOf(codes == null ? BASE : codes);
    }
}

File: erp-frontend/src/app/shared/constants/consigne-codes.ts
/**
 * Codes produits consigne (emballages consignés) — identiques au module Odoo blessing_consulting.
 * Source unique : avant cette constante, 5 fichiers dupliquaient chacun leur propre liste, et
 * sales/orders/order-form.component.ts avait dérivé (19 codes au lieu de 38) — une commande vente
 * avec un emballage manquant n'était pas traitée en consigne à la saisie alors que la facture
 * générée à partir de la même commande la traitait bien comme consigne.
 */
export const CONSIGNE_NAMES: Record<string, string> = {
  CB12:   'CASIER BOUTEILLE 12',
  CB24:   'CASIER BOUTEILLE 24',
  CB12M:  'CASIER BOUTEILLE 12 (METAL)',
  CB24M:  'CASIER BOUTEILLE 24 (METAL)',
  CV12:   'CASIER VERRE 12',
  CV24:   'CASIER VERRE 24',
  CBG12:  'CASIER BOUTEILLE GUINNESS 12',
  CBG15:  'CASIER BOUTEILLE GUINNESS 15',
  CBG24:  'CASIER BOUTEILLE GUINNESS 24',
  CVG12:  'CASIER VERRE GUINNESS 12',
  CVG15:  'CASIER VERRE GUINNESS 15',
  CVG24:  'CASIER VERRE GUINNESS 24',
  VIP12:  'VIP 12',
  VIP24:  'VIP 24',
  VCP12:  'VCP 12',
  VCP24:  'VCP 24',
  VIPG12: 'VIP GUINNESS 12',
  VIPG15: 'VIP GUINNESS 15',
  VIPG24: 'VIP GUINNESS 24',
  EGUI12: 'EMBALLAGE GUINNESS 12',
  EGUI15: 'EMBALLAGE GUINNESS 15',
  EGUI24: 'EMBALLAGE GUINNESS 24',
  PP:     'PALETTE PLASTIQUE',
  PB:     'PALETTE BOIS',
  TT:     'TONNELET',
  BPM:    'BOUTEILLE PET METAL',
  BGM:    'BOUTEILLE GUINNESS METAL',
  CAIMET: 'CAISSE METALLIQUE',
  CONS001: 'CONSIGNE DIVERSE',
  INPN33: 'EMBALLAGE INPN 33',
  EMB1:   'EMBALLAGE 1',
  EMB2:   'EMBALLAGE 2',
  EMB3:   'EMBALLAGE 3',
  EMB4:   'EMBALLAGE 4',
  EMB5:   'EMBALLAGE 5',
  CAISMB: 'CAISSE METAL BOUTEILLE',
  'PALT-V': 'PALETTE VERRE',
  PALTPL: 'PALETTE PLASTIQUE (PL)',
  PRC01:  'PORTE-CASIER 01',
  ELV01:  'ELEVATEUR 01',
};

export const CONSIGNE_CODES: Set<string> = new Set(Object.keys(CONSIGNE_NAMES));
